src/scheduler.ml

```diff
@@ -2,38 +2,116 @@ open! Stdune
 open Import
 open Fiber.O
 
-type status_line_config =
-  { message   : string option
-  ; show_jobs : bool
-  }
-
-type running_job =
+type job =
   { pid  : int
   ; ivar : Unix.process_status Fiber.Ivar.t
   }
 
-module Watch : sig
-  (** Runs forever, communicating events to channel when changes are
-      detected. *)
-  val wait_to_chan : Path.t list Event.channel -> 'a
+(** The event queue *)
+module Event : sig
+  type t =
+    | Files_changed
+    | Job_completed of job * Unix.process_status
+
+  (** Return the next event. File changes event are always flattened
+      and returned first. *)
+  val next : unit -> t
 
-  (** Compares a batch of changed files to ignored files list.
+  (** Ignore the ne next file change event about this file. *)
+  val ignore_next_file_change_event : Path.t -> unit
 
-      This should be only called from the main thread. *)
-  val should_trigger_rebuild : Path.t list -> bool
+  (** Register the fact that a job was started. *)
+  val register_job_started : unit -> unit
 
-  (** Ignore the next event about this file. *)
-  val ignore_file : Path.t -> unit
+  (** Number of jobs for which the status hasn't been reported yet .*)
+  val pending_jobs : unit -> int
+
+  (** Send an event to the main thread. *)
+  val send_files_changed : Path.t list -> unit
+  val send_job_completed : job -> Unix.process_status -> unit
 end = struct
+  type t =
+    | Files_changed
+    | Job_completed of job * Unix.process_status
+
+  let jobs_completed = Queue.create ()
+  let files_changed = ref []
+  let mutex = Mutex.create ()
+  let cond = Condition.create ()
+
   let ignored_files = String.Table.create 64
```

Comment:
Can this be a table of paths instead?

Reply:
Also, a comment for this table would be nice as well. I.e. something like "stores the paths of the file events to ignore in the next event batch"

Reply:
The issue is that fswatch/inotifyway are reporting absolute filenames, while we register files to ignore that are paths in the source tree.